share/wake/lib/core/json.wake

```diff
@@ -169,3 +169,143 @@ export def x ==/ y = match x y
   (JArray a) _ = exists (_ ==/ y) a
   _ (JArray b) = exists (_ ==/ x) b
   _ _ = False
+
+# Simplify a JSON structure according to the rules given.
+# For example, Wake's implementation of JSON is more lenient than the base
+# standards, so if maximum compatibility is desired in situations where some
+# other implementation may be parsing generated output, `normalizeJSONCompat`
+# may be used to target that lowest common denominator.
+#
+# Note that the relevant normalization is applied to each member of a `JArray`
+# or `JObject` *before* the enclosing container. This means that if some
+# normalization fails anywhere in the JSON tree the entire call will fail, even
+# if the specific branch causing the failure would later have been pruned.
+export def normalizeJSONWith (fmt: JSONNormalize) (json: JValue): Result JValue Error =
+    def normalizeJSON (path: String) = match _
+        JString a =
+            getJSONNormalizeString fmt a
+            | rmap JString
+            | addErrorContext path
+        JInteger a =
+            getJSONNormalizeInteger fmt a
+            | rmap JInteger
+            | addErrorContext path
+        JDouble a =
+            getJSONNormalizeDouble fmt a
+            | rmap JDouble
+            | addErrorContext path
+        JBoolean a =
+            getJSONNormalizeBoolean fmt a
+            | rmap JBoolean
+            | addErrorContext path
+        JNull =
+            Pass JNull
+        JObject a =
+            def normalizeJObjectValue entry =
+                def Pair key value =
+                    entry
+                normalizeJSON "{path}/{key}" value
```

Comment:
So, what I was proposing is that you first combine the keys shallowly, using 'JSONNormalizeObject'. Then, you recurse on the combine list.

Reply:
I'm pretty concerned that your current algorithm takes `O(n^2)` time a JSON with nesting `n` levels deep.

Reply:
That is definitely bad, and I will then need to rework the `deduplicateJObjectKeys`.  Applying the `normalizeJSONWith` recursion *after* joining the keys might indeed wind up being the only way to avoid it, but the reason I'm balking at that is because doing so means that if a user writes a so-called "normalize" function to make every int 0, I think they'd expect that every int in an object should be able to be collapsed rather than an error being thrown on what will become identical.  I'm pretty sure I can make `deduplicateJObjectKeys` recurse only to `JObjects` that are currently being joined due to duplication; that would still be `O(n^2)` at worst, but the most common case (without duplicate keys) would be `O(n)` or thereabouts, and even a duplicate key every other level wouldn't be too much more complex.  Would that be acceptable?

Reply:
The more realistic use case would be string canonicalization: single-pass recursion purely through `normalizeJSONWith` -- i.e. children after the current node -- couldn't catch `{"eacute":"é", "eacute":"é"}` and would fail to deduplicate those.  Normalizing children first would collapse the second (with a combining acute) into a string identical to the first (precomposed), and when it then looks at the enclosing object, would see two duplicate keys with identical values and happily return `{"eacute":"é"}`.